lib/buffer.js

```diff
@@ -234,34 +237,29 @@ function copyImpl(source, target, targetStart, sourceStart, sourceEnd) {
     if (sourceEnd < 0)
       throw new ERR_OUT_OF_RANGE('sourceEnd', '>= 0', sourceEnd);
   }
+  // Guaranteed 0 <= sourceEnd
 
   if (targetStart >= target.byteLength || sourceStart >= sourceEnd)
     return 0;
 
+  // Guaranteed 0 <= sourceStart < sourceEnd (NO GUARANTEE <= source.byteLnegth)
+  // Guaranteed 0 <= targetStart < target.byteLength here
+
   return _copyActual(source, target, targetStart, sourceStart, sourceEnd);
 }
 
+// Assumes `source`, `target` are real buffers.
+// Assumes 0 <= sourceStart <= sourceEnd
+// Assumes 0 <= targetStart <= target.byteLength
```

Comment:
I think we could just make these asserts. The perf impact is negligable...

Reply:
https://github.com/nodejs/node/blob/209289ceb27c660c065210308784debc58863c25/lib/buffer.js#L271-L275